src/services/codefixes/fixUnusedIdentifier.ts

```diff
@@ -334,6 +334,34 @@ namespace ts.codefix {
         }
     }
 
+    function mayDeleteParameter(p: ParameterDeclaration, checker: TypeChecker) {
+        const parent = p.parent;
+        switch (parent.kind) {
+            case SyntaxKind.Constructor:
+            case SyntaxKind.FunctionDeclaration:
+                return true;
+
+            case SyntaxKind.FunctionExpression:
+            case SyntaxKind.ArrowFunction:
+                // Can't remove a non-last parameter in a callback. (Can if future parameters are also unused.)
+                const index = parent.parameters.indexOf(p);
+                Debug.assert(index !== -1);
+                return parent.parameters.slice(index).every(p => p.name.kind === SyntaxKind.Identifier && !p.symbol.isReferenced) || !checker.getContextualType(parent);

```

Comment:
even if you did not have a contextual type, the function may be called or assigned to something else that passes or expects different parameters.

Reply:
That's true, though it would mean we could never remove any parameters. Maybe we should play it safe in code fix all but still provide the option to delete as a single fix?

Reply:
I think it is always safe to remove from the end, contextual type or not. 

We should think about updating call sites to remove the unused arguments as well.. but that is a different change..

Reply:
So never remove a non-last parameter? A method in a class may also be expected to conform to some interface that wasn't declared.

Reply:
a method with fewer required arguments is assignable to one with more. so that should work. but yes, it could result in errors.. really any change to the signature can result in errors.. so we can either be conservative and say we only do it if we know all references, and know it was never aliased, or we can be more open and say, as long as the error will be caught by type check, we should be fine..